File: Week_02/add_digits/Solution.java
package add_digits;

/**
 * 循环
 * 时间复杂度：O(n)
 * 空间复杂度：O(1)
 */
class Solution1 {
    public int addDigits(int num) {
        while (num >= 10) {
            num = (num / 10) + (num % 10);
        }
        return num;
    }
}

/**
 * 递归
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */
class Solution2 {
    public int addDigits(int num) {
        if (num < 10) return num;
        num = (num / 10) + (num % 10);
        return addDigits(num);
    }
}

/**
 * 数学推导
 * 时间复杂度：O(1)
 * 空间复杂度：O(1)
 */
class Solution3 {
    public int addDigits(int num) {
        return (num - 1) % 9 + 1;
    }
}

File: Week_02/add_digits/Solution.py
class Solution1:
    def addDigits(self, num: int) -> int:
        while num >= 10:
            num = int(num / 10) + num % 10
        return num


class Solution2:
    def addDigits(self, num: int) -> int:
        if num < 10: return num
        return self.addDigits(int(num / 10) + num % 10)


class Solution3:
    def addDigits(self, num: int) -> int:
        return 0 if num == 0 else (num - 1) % 9 + 1